### PR TITLE
feature: Verbatim boxes

### PR DESCRIPTION
**Pull request by speller26**
Introduces verbatim boxes, which allows users to tell the compiler not
to modify specified parts of the circuit.

*Issue #, if available:*

*Description of changes:*

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-sdk-python/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by revendar — APPROVED** [review]

File: src/braket/circuits/circuit.py [hunk 1]
```diff
@@ -487,6 +492,77 @@ def add_circuit(
...
+        if verbatim_circuit.result_types:
+            raise ValueError("Verbatim subcircuit is not measured and cannot have result types")
+
+        if verbatim_circuit.instructions:
+            self.add_instruction(Instruction(compiler_directives.StartVerbatimBox()))
+            for instruction in verbatim_circuit.instructions:
+                self.add_instruction(instruction, target_mapping=target_mapping)
+            self.add_instruction(Instruction(compiler_directives.EndVerbatimBox()))
+
+        self._has_compiler_directives = True
```
Comment (kshitijc) [inline src/braket/circuits/circuit.py:563]: If verbatim_circuit is empty then should self._has_compiler_directives be True ?

**Review by jheckey — DISMISSED** [review]

File: src/braket/circuits/circuit.py [hunk 2]
```diff
@@ -487,6 +492,77 @@ def add_circuit(
...
+             Instruction('operator': H('qubit_count': 1), 'target': QubitSet([Qubit(11)])),
+             Instruction('operator': EndVerbatimBlock, 'target': QubitSet([]))]
+
+            >>> widget = Circuit().h(0).cnot(0, 1)
+            >>> circ = Circuit().add_verbatim_box(widget, target=[10, 11])
+            >>> print(list(circ.instructions))
+            [Instruction('operator': StartVerbatimBlock, 'target': QubitSet([])),
+             Instruction('operator': H('qubit_count': 1), 'target': QubitSet([Qubit(10)])),
+             Instruction('operator': H('qubit_count': 1), 'target': QubitSet([Qubit(11)])),
+             Instruction('operator': EndVerbatimBlock, 'target': QubitSet([]))]
```
Comment (kshitijc) [inline src/braket/circuits/circuit.py:545]: Block -> Box

File: src/braket/circuits/circuit.py [hunk 3]
```diff
@@ -487,6 +492,77 @@ def add_circuit(
...
+        """
+        if target_mapping and target is not None:
```
Comment (kshitijc) [inline src/braket/circuits/circuit.py:547]: Do we need to support both target and target_mapping at launch ? Was this a requirement ?

File: src/braket/circuits/circuit.py [hunk 4]
```diff
@@ -900,6 +978,16 @@ def as_unitary(self) -> np.ndarray:
...
 
+    @property
+    def qubits_frozen(self) -> bool:
+        """bool: Whether the circuit's qubits are frozen, that is, cannot be remapped.
+
+        This may happen because the circuit contains compiler directives preventing compilation
+        of a part of the circuit, which consequently means that none of the other qubits can be
+        rewired either for the program to still make sense.
+        """
+        return self._has_compiler_directives
```
Comment (kshitijc) [inline src/braket/circuits/circuit.py:989]: Does every compiler directive imply frozen qubits?

File: test/unit_tests/braket/circuits/test_ascii_circuit_diagram.py [hunk 5]
```diff
@@ -205,6 +205,160 @@ def test_connector_across_non_used_qubits():
...
+        "      |               | |             ",
+        "q1 : -*************---X-***********---",
+        "",
+        "T  : |      0      |1|2|     3     |4|",
+    )
+    expected = "\n".join(expected)
+    assert AsciiCircuitDiagram.build_diagram(circ) == expected
+
+
+def test_verbatim_3q_no_preceding():
```
Comment (kshitijc) [inline test/unit_tests/braket/circuits/test_ascii_circuit_diagram.py:292]: Could we add a test with verbatim box applied to a subset of qubits in the circuit ?

File: src/braket/circuits/circuit.py (hunk 1 above)
Comment (speller26) [reply src/braket/circuits/circuit.py:563]: Fixed

File: src/braket/circuits/circuit.py (hunk 4 above)
Comment (speller26) [reply src/braket/circuits/circuit.py:989]: No, and that's why `_has_compiler_directives` is private; the actual check happening under the hood can change in the future.

File: src/braket/circuits/circuit.py (hunk 3 above)
Comment (speller26) [reply src/braket/circuits/circuit.py:547]: Wasn't a requirement, and I can remove this. However, the reason I made this was for consistency with the rest of the `add_` methods

File: test/unit_tests/braket/circuits/test_ascii_circuit_diagram.py (hunk 5 above)
Comment (speller26) [reply test/unit_tests/braket/circuits/test_ascii_circuit_diagram.py:292]: Added!

**Review by revendar — APPROVED** [review]

**Pull request merged**